server/oauth_migration.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-api/cluster"
+	"github.com/pkg/errors"
+)
+
+const (
+	oauthMigrationStoreKey = "oauth_migration"
+	oauthMigrationMutexKey = "oauth_migration_mutex"
+
+	oauthMigrationStatusInProgress = "IN_PROGRESS"
+	oauthMigrationStatusComplete   = "COMPLETE"
+	oauthMigrationStatusError      = "ERROR"
+)
+
+func (p *Plugin) checkAndPerformOAuthTokenMigration() error {
+	mutexAPI := cluster.MutexPluginAPI(p.API)
+	mutex, err := cluster.NewMutex(mutexAPI, oauthMigrationMutexKey)
+	if err != nil {
+		return err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
```

Comment:
In the first commit of the PR, I had this call further down, and not `defer`ed. I think it should be `defer`ed up here so we can make sure the migration completes before any of the plugins are considered "activated". This way we can avoid the thundering hurd problem of several browsers connecting at once.

Reply:
Note that this means all plugin instances are blocking on `OnActivate` until the migration is complete. This will slow down the start-up time of the server. I wonder how we can make this more asynchronous. We have control over the frontend of the plugin so we can make changes there if necessary.